apps/dashboard/src/contract-ui/hooks/useContractSources.ts

```diff
@@ -1,17 +1,48 @@
-import { useQueryWithNetwork } from "@3rdweb-sdk/react/hooks/query/useQueryWithNetwork";
-import { useSDK } from "@thirdweb-dev/react";
+import { thirdwebClient } from "@/constants/client";
+import { useQuery } from "@tanstack/react-query";
+import type { ThirdwebContract } from "thirdweb";
+import { getCompilerMetadata } from "thirdweb/contract";
+import { download } from "thirdweb/storage";
+import invariant from "tiny-invariant";
 
-export function useContractSources(contractAddress?: string) {
-  const sdk = useSDK();
-  return useQueryWithNetwork(
-    ["contract-sources", contractAddress],
-    async () => {
-      return sdk
-        ?.getPublisher()
-        .fetchContractSourcesFromAddress(contractAddress || "");
-    },
-    {
-      enabled: !!contractAddress,
+// An example for a contract that has IPFS URIs in its metadata: abstract-testnet/0x8A24a7Df38fA5fCCcFD1259e90Fb6996fDdfcADa
+
+export function useContractSources(contract?: ThirdwebContract) {
+  return useQuery({
+    queryKey: [
+      "contract-sources",
+      contract?.chain.id || "",
+      contract?.address || "",
+    ],
+    queryFn: async (): Promise<Array<{ filename: string; source: string }>> => {
+      invariant(contract, "contract is required");
+      const data = await getCompilerMetadata(contract);
+      const sources = data.metadata.sources || {};
+      return await Promise.all(
+        // biome-ignore lint/suspicious/noExplicitAny: <explanation>
+        Object.entries(sources).map(async ([path, info]: [string, any]) => {
+          const urls = info.urls as string[];
```

Comment:
too many `as <x>` here, please type this properly, it seems like `getCompilerMetadata` should be typed correctly -> that should solve things?

Reply:
It's because we have `any` in the return type from the SDK:
```
export type PublishedMetadata = {
  name: string;
  abi: Abi;
  metadata: Record<string, any>; // <-------------------
  info: {
    title?: string;
    author?: string;
    details?: string;
    notice?: string;
  };
  licenses: string[];
  isPartialAbi?: boolean;
};
```

How/where can I get the type for this?

Reply:
probably need to type it correctly inside SDK then